akka-stream/src/main/scala/akka/stream/impl/fusing/FutureFlow.scala

```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright (C) 2020 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.stream.impl.fusing
+
+import akka.annotation.InternalApi
+import akka.dispatch.ExecutionContexts
+import akka.stream.{ Attributes, FlowShape, Inlet, NeverMaterializedException, Outlet }
+import akka.stream.scaladsl.{ Flow, Keep, Source }
+import akka.stream.stage.{ GraphStageLogic, GraphStageWithMaterializedValue, InHandler, OutHandler }
+import akka.util.OptionVal
+
+import scala.concurrent.{ Future, Promise }
+import scala.util.{ Failure, Success, Try }
+
+@InternalApi private[akka] final class FutureFlow[In, Out, M](future: Future[Flow[In, Out, M]])
+    extends GraphStageWithMaterializedValue[FlowShape[In, Out], Future[M]] {
+  val in = Inlet[In](s"${this}.in")
+  val out = Outlet[Out](s"${this}.out")
+  override val shape: FlowShape[In, Out] = FlowShape(in, out)
+
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes): (GraphStageLogic, Future[M]) = {
+    val pr = Promise[M]
+    val logic = new GraphStageLogic(shape) {
+
+      //seems like we must set handlers BEFORE preStart
+      setHandlers(in, out, initializing)
+
+      override def preStart(): Unit = {
+        super.preStart()
+        future.value match {
+          case Some(tr) =>
+            initializing.onFuture(tr)
+          case None =>
+            val cb = getAsyncCallback(initializing.onFuture)
+            future.onComplete(cb.invoke)(ExecutionContexts.parasitic)
+            //in case both ports are closed before future completion
+            setKeepGoing(true)
+        }
+      }
+
+      override def postStop(): Unit = {
+        if (!pr.isCompleted) {
+          pr.failure(new NeverMaterializedException())
+        }
+        super.postStop()
+      }
+
+      object initializing extends InHandler with OutHandler {
+        override def onPush(): Unit = {
+          //just leave it in the port
```

Comment:
good point, if I remember the pull instead of actually forwarding it I wouldn't have to buffer elements or rely on the port to keep it for me.
b.t.w my impl assumes the element is kept in the port as long as it's not pulled again, what happens in case of an upstream error? which btw may happen out of band... I guess this will remove the element from the port, right?
This also relates to some conversation I had with @johanandren in [this comment](https://github.com/akka/akka/issues/28016#issuecomment-595059713) on #28204 
so if we never pull upstream before the future flow was materialized we don't have to mess with this, right? if an out of band failure happens we won't have any pending upstream pull hence we definitely won't have a pending element in the port.

how do you suggest `flatMapPrefix` to handle this? we must pull upstream there, and we explicitly buffer pushed elements, but currently in case of an 'early' upstream failure we don't materialize the flow. this theoretically prevents users from providing flows with 'recover' operators (though they will handle 'late' failures)

Reply:
> so if we never pull upstream before the future flow was materialized we don't have to mess with this, right? if an out of band failure happens we won't have any pending upstream pull hence we definitely won't have a pending element in the port.

Yes, failures will clear the port which usually is fine since you the failure should be propagated as quick as possible.

Reply:
> how do you suggest `flatMapPrefix` to handle this? we must pull upstream there, and we explicitly buffer pushed elements, but currently in case of an 'early' upstream failure we don't materialize the flow. this theoretically prevents users from providing flows with 'recover' operators (though they will handle 'late' failures)

I see, that's a bit of a weird edge case. If we wanted to give `flatMapPrefix` users more control we would then have to think about how to represent the case where only part of the prefix has been collected and then the stream failed. Not sure we want to go into that right now? Maybe we wait until someone asks for that feature.

Reply:
I think a custom attribute is the way to go, but probably not now 😎